server/streamable_http.go

```diff
@@ -0,0 +1,812 @@
+	// Get session ID from header if present
+	sessionID := r.Header.Get("Mcp-Session-Id")
+	var session *streamableHTTPSession
+
+	// Check if this is a request with a valid session
+	if sessionID != "" {
+		if sessionValue, ok := s.sessions.Load(sessionID); ok {
+			if sess, ok := sessionValue.(SessionWithTools); ok {
+				session = sess.(*streamableHTTPSession)
+			} else {
+				http.Error(w, "Invalid session", http.StatusBadRequest)
+				return
+			}
+		} else {
+			// Session not found
+			http.Error(w, "Session not found", http.StatusNotFound)
+			return
+		}
+	}
+
+	// Parse the request body
+	var rawMessage json.RawMessage
+	if err := json.NewDecoder(r.Body).Decode(&rawMessage); err != nil {
+		http.Error(w, "Invalid JSON", http.StatusBadRequest)
+		return
+	}
+
+	// Parse the base message to determine if it's a request or notification
+	var baseMessage struct {
+		JSONRPC string      `json:"jsonrpc"`
+		Method  string      `json:"method"`
+		ID      interface{} `json:"id,omitempty"`
+	}
+	if err := json.Unmarshal(rawMessage, &baseMessage); err != nil {
+		http.Error(w, "Invalid JSON-RPC message", http.StatusBadRequest)
+		return
+	}
+
+	// Create context for the request
+	ctx := r.Context()
+	if session != nil {
+		ctx = s.server.WithContext(ctx, session)
+	}
+	if s.contextFunc != nil {
+		ctx = s.contextFunc(ctx, r)
+	}
+
+	// Handle the message based on whether it's a request or notification
+	if baseMessage.ID == nil {
+		// It's a notification
+		s.handleNotification(w, ctx, rawMessage)
+	} else {
+		// It's a request
+		s.handleRequest(w, r, ctx, rawMessage, session)
+	}
+}
+
+// handleNotification processes JSON-RPC notifications
+func (s *StreamableHTTPServer) handleNotification(w http.ResponseWriter, ctx context.Context, rawMessage json.RawMessage) {
+	// Process the notification
+	s.server.HandleMessage(ctx, rawMessage)
+
+	// Return 202 Accepted for notifications
+	w.WriteHeader(http.StatusAccepted)
+}
+
+// handleRequest processes JSON-RPC requests
+func (s *StreamableHTTPServer) handleRequest(w http.ResponseWriter, r *http.Request, ctx context.Context, rawMessage json.RawMessage, session *streamableHTTPSession) {
+	// Parse the request to get the method and ID
+	var request struct {
+		JSONRPC string      `json:"jsonrpc"`
+		Method  string      `json:"method"`
+		ID      interface{} `json:"id"`
+	}
+	if err := json.Unmarshal(rawMessage, &request); err != nil {
+		http.Error(w, "Invalid JSON-RPC request", http.StatusBadRequest)
+		return
+	}
+
+	// Check if this is an initialization request
+	isInitialize := request.Method == "initialize"
+
+	// If this is not an initialization request and we don't have a session,
+	// and we're not in stateless mode, then reject the request
+	if !isInitialize && session == nil && !s.statelessMode {
+		http.Error(w, "Bad Request: Server not initialized", http.StatusBadRequest)
+		return
+	}
+
+	// Process the request
+	response := s.server.HandleMessage(ctx, rawMessage)
```

Comment:
I think there's no handling for server -> client notifications before mcp server response.